### PR TITLE
Convert de.zalando.zally.apireview to Kotlin

### LINKED ISSUE #712: Convert remaining server code from Java to Kotlin
In the OpenAPI 3 PR I [commented surprise](https://github.com/zalando/zally/pull/706/files/59e4895efa750befefe125a8ef55694aa5396227#r188894438) at adding more Java code to the codebase but figured I should elevate discussion. We've fully embraced Kotlin as a language choice for the rules, a few supporting classes, and many of the tests. There is a little (actually surprisingly little) cognitive dissonance mentally switching languages when switching between parts of the code, so I wonder if now is the time to fully embrace Kotlin and switch the rest of the server code over.

Are there good reasons not to make this switch?

### PR DESCRIPTION
**Pull request by roxspring**
- Converted `de.zalando.zally.apireview` to Kotlin via Intellij Idea
- Applied Kotlin idioms
- Completes the transition to pure Kotlin codebase

Closes #712

### CODE REVIEW

**Comment by roxspring** [comment]
Any chance of review today? Would be great to avoid the branches getting stale while I disappear for a week!

@ddufourboivin-zalando, @maxim-tschumak, @netme and @tkrop

**Comment by tkrop** [comment]
:+1:

File: server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.springframework.http.HttpStatus
+import org.springframework.http.HttpStatus.UNSUPPORTED_MEDIA_TYPE
+import org.springframework.http.MediaType.parseMediaTypes
+import org.springframework.stereotype.Component
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.ResourceAccessException
+import org.springframework.web.client.RestTemplate
+
+@Component
+class ApiDefinitionReader @Autowired
```
[the remark this answers is not in the document]
Comment (ddufourboivin-zalando) [reply server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt]: As discussed with @tkrop , I will propose a "better Kotlin" pull-request after this is merged.

File: server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+constructor(private val client: RestTemplate) {
+
+    fun read(request: ApiDefinitionRequest): String? = when {
```
[the remark this answers is not in the document]
Comment (ddufourboivin-zalando) [reply server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt]: As discussed with @tkrop , I will propose a "better Kotlin" pull-request after this is merged.

File: server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        else -> throw MissingApiDefinitionException()
+    }
+
+    private fun readFromUrl(url: String?): String? = try {
+        val headers = HttpHeaders()
+        headers.accept = MEDIA_TYPE_WHITELIST
+
+        val entity = HttpEntity<String>(null, headers)
+
+        val response = client.exchange(removeSpecialCharactersSuffix(url!!), HttpMethod.GET, entity, String::class.java)
```
[the remark this answers is not in the document]
Comment (ddufourboivin-zalando) [reply server/src/main/java/de/zalando/zally/apireview/ApiDefinitionReader.kt]: As discussed with @tkrop , I will propose a "better Kotlin" pull-request after this is merged.

**Comment by ddufourboivin-zalando** [comment]
👍

**Comment by ddufourboivin-zalando** [comment]
@roxspring : The build is failing because of `ktlint` violations.
Please run `./gradlew clean ktlint` locally and fix them.
Before committing (or opening a PR), make sure you run all checks (tests, linting, etc...) by running `./gradlew clean check`.

File: build-and-run.sh [hunk 4]
```diff
@@ -19,7 +19,7 @@ cd ${SERVER_DIR} && ./gradlew clean build
 # Build web ui
 cd ${WEB_UI_DIR} && yarn && yarn build
 
-if [ "$@" == "--bark" ]; then
+if [[ "$@" == "--bark" ]]; then
```
Comment (roxspring) [inline build-and-run.sh]: Meh. Hadn't intended to commit this here... but having just started to use `build-and-run.sh` under Windows Subsystem for Linux with Ubuntu 18.04, the change appears to be necessary. Haven't checked it under pure Linux or OSX yet. Can back out if necessary.

File: server/src/main/java/de/zalando/zally/apireview/ApiReview.kt [hunk 5]
```diff
@@ -89,20 +89,20 @@ class ApiReview : Serializable {
...
+                apiDefinition == that.apiDefinition &&
+                isSuccessfulProcessed == that.isSuccessfulProcessed &&
+                day == that.day &&
+                created == that.created &&
+                numberOfEndpoints == that.numberOfEndpoints &&
+                mustViolations == that.mustViolations &&
+                shouldViolations == that.shouldViolations &&
+                mayViolations == that.mayViolations &&
+                hintViolations == that.hintViolations &&
+                ruleViolations == that.ruleViolations)
```
Comment (roxspring) [inline server/src/main/java/de/zalando/zally/apireview/ApiReview.kt]: Ideally `ApiReview` and other DTO classes would become Kotlin `data class` and not need custom `hashCode()` or `equals()`. Didn't manage to make that transition in this PR but figure that can be improved later.

**Comment by maxim-tschumak** [comment]
:+1:

**Pull request merged**